KAFKA-6362: maybeAutoCommitOffsetsAsync should try to discover coordinator

huxihx opened the pull request:
Currently, `maybeAutoCommitOffsetsAsync` may not retry to find out coordinator even after the coordinator goes back to service. As a result, all asynchronous offset commits will fail.

This patch refines `maybeAutoCommitOffsetsAsync` to have it periodically retry the coordinator discovering.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

huxi-2b committed:
KAFKA-6362: ConsumerCoordinator.maybeAutoCommitOffsetsAsync should try to discover coordinator.

Currently, `maybeAutoCommitOffsetsAsync` may not retry to find out coordinator even after the coordinator goes back to service. As a result, all asynchronous offset commits will fail.

This patch refines `maybeAutoCommitOffsetsAsync` to have it periodically retry the coordinator discovering.

huxihx commented:
@hachikuji Please take some time to review it. Thanks.

huxihx commented:
Could alternatively do the finding job only for user-assigned consumers.

huxihx commented:
@hachikuji Could you  take some time to review this minor patch? Thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -624,10 +624,8 @@ public boolean commitOffsetsSync(Map<TopicPartition, OffsetAndMetadata> offsets,
-            } else if (now >= nextAutoCommitDeadline) {
-                this.nextAutoCommitDeadline = now + autoCommitIntervalMs;
+            if (now >= nextAutoCommitDeadline) {
+                this.nextAutoCommitDeadline = now + (coordinatorUnknown() ? retryBackoffMs : autoCommitIntervalMs);
```
I think a better approach might be to set `nextAutoCommitDeadline` in the offset commit callback in `doAutoCommitOffsetsAsync`. If there was an error, we can set it based on the retry backoff; otherwise we can use the auto commit interval.

hachikuji reviewed (commented):
Thanks for the patch. Left one comment. Also we will need a test case.

hachikuji commented:
@huxihx One additional suggestion: inside `ConsumerCoordinator.maybeAutoCommitOffsetsNow`, we should probably remove the check for `coordinatorUnknown()`.

huxi-2b committed:
Merge branch 'trunk' of https://github.com/apache/kafka into KAFKA-6362

huxi-2b committed:
KAFKA-6362: maybeAutoCommitOffsetsAsync should try to discover coordinator

1. Added a test case to cover this scenario
2. Remove check for `coordinatorUnknown` in `ConsumerCoordinator.maybeAutoCommitOffsets`
3. Merge `maybeAutoCommitOffsetsNow` with `maybeAutoCommitOffsetsAsync`
4. Make `maybeAutoCommitOffsetsAsync` public

huxihx commented:
@hachikuji  Thanks for the comments. Please review the new code with changes below:
1. Added a test case to cover this scenario
2. Remove check for `coordinatorUnknown` in `ConsumerCoordinator.maybeAutoCommitOffsets`
3. Merge `maybeAutoCommitOffsetsNow` with `maybeAutoCommitOffsetsAsync`
4. Make `maybeAutoCommitOffsetsAsync` public

huxihx commented:
@junrao Since 1.1 release is around the corner, could you please take some time to review the latest patch? Thank you:)

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/286/

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -623,24 +623,15 @@ public boolean commitOffsetsSync(Map<TopicPartition, OffsetAndMetadata> offsets,
-
-    private void doAutoCommitOffsetsAsync() {
-        Map<TopicPartition, OffsetAndMetadata> allConsumedOffsets = subscriptions.allConsumed();
+    private void doAutoCommitOffsetsAsync(Map<TopicPartition, OffsetAndMetadata> allConsumedOffsets) {
```
Why add this argument? As far as I can tell, all uses just pass `subscriptions.allConsumed()`, so the original was fine?

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -1625,6 +1626,34 @@ public void testHeartbeatThreadClose() throws Exception {
     }
 
+    @Test
+    public void testAsyncOffsetCommitAfterCoordinatorBackToService() {
```
This test case passes without the fix. It doesn't look like it even goes through the auto-commit path.

hachikuji replied:
The name should mention the fact that this case covers auto-commit.

hachikuji reviewed (commented):
Thanks for the updates. Left a couple comments.

huxi-2b committed:
KAFKA-6362: Address Jason's comments, including:

1. Restore the method signature for `maybeAutoCommitOffsetsAsync` and `doAutoCommitOffsetsAsync`
2. Refine `ConsumerCoordinatorTest.testAsyncOffsetCommitAfterCoordinatorBackToService` to go through the auto-commit path
3. In `ConsumerCoordinator.commitOffsetsAsync`, add `client.pollNoWakeup();` in lookupCoordinator's onSuccess to have it commit offsets immediately after finding out coordinator

huxihx commented:
@hachikuji Thanks for the comments. Submitted a new patch including changes below:

1. Restore the method signature for `maybeAutoCommitOffsetsAsync` and `doAutoCommitOffsetsAsync`
2. Refine `ConsumerCoordinatorTest.testAsyncOffsetCommitAfterCoordinatorBackToService` to go through the auto-commit path
3. In `ConsumerCoordinator.commitOffsetsAsync`, add `client.pollNoWakeup();` in lookupCoordinator's onSuccess to have it commit offsets immediately after finding out coordinator

Please review again. Thanks.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/367/

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -623,20 +624,11 @@ public boolean commitOffsetsSync(Map<TopicPartition, OffsetAndMetadata> offsets,
-
-    public void maybeAutoCommitOffsetsNow() {
-        if (autoCommitEnabled && !coordinatorUnknown())
+    // visible only for testing
```
This is called from `KafkaConsumer`, so it is not just visible for testing.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -98,13 +98,12 @@
     private MockTime time;
-    private MockClient client;
-    private Cluster cluster = TestUtils.clusterWith(1, new HashMap<String, Integer>() {
+    private MockClient client;    private Cluster cluster = TestUtils.clusterWith(1, new HashMap<String, Integer>() {
```
This looks unintentional.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -98,13 +98,12 @@
     private int heartbeatIntervalMs = 5000;
     private long retryBackoffMs = 100;
-    private boolean autoCommitEnabled = false;
+    private boolean autoCommitEnabled = true;
```
We intentionally disable auto-commit so that it does not interfere with other test cases. If we are testing auto-commit behavior in a test, we just enable it locally for that test case (see `testAutoCommitDynamicAssignment` for example). Let's continue to follow this pattern.

hachikuji reviewed (commented):
Thanks for the updates. Left a few more comments.

huxi-2b committed:
address Jason's comments including
1. Remove incorrect comment for `maybeAutoCommitOffsetsAsync`
2. Rename test case name to `testAutoCommitAfterCoordinatorBackToService`
3. Use local coordinator

huxihx commented:
@hachikuji Thanks for the comments and correct me if I am wrong. "This test case passes without the fix. It doesn't look like it even goes through the auto-commit path." -- Seems the test got failed without this fix now. Do you think it's a valid test case?

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -1625,6 +1625,25 @@ public void testHeartbeatThreadClose() throws Exception {
+        client.prepareResponse(offsetCommitResponse(Collections.singletonMap(t1p, Errors.NONE)));
+
+        // Async commit offset should find coordinator
+        coordinator.maybeAutoCommitOffsetsAsync(time.nanoseconds() + autoCommitIntervalMs); //make sure it does happen
```
I think `nanoseconds` should be `milliseconds`. It's a little more idiomatic to do it like this:

```java
time.sleep(autoCommitIntervalMs);        
coordinator.maybeAutoCommitOffsetsAsync(time.milliseconds());
```

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -1625,6 +1625,25 @@ public void testHeartbeatThreadClose() throws Exception {
+        ConsumerCoordinator coordinator = buildCoordinator(new Metrics(), assignors,
+                ConsumerConfig.DEFAULT_EXCLUDE_INTERNAL_TOPICS, true, true);
+        subscriptions.assignFromUser(Collections.singleton(t1p));
+        subscriptions.committed(t1p, new OffsetAndMetadata(0));
```
This is not actually needed.

hachikuji reviewed (commented):
Thanks for the update. I think the new test case is reasonable. I tried it without the fix and it fails as expected. I left a couple additional comments.

huxi-2b committed:
Removed useless subscriptions.committed() call and took a more reasonable sleep call to ensure auto commit does happen.

huxihx commented:
@hachikuji Thanks for the comments.  Please review again.

hachikuji reviewed (approved):
LGTM. Thanks for the patch. I will merge to trunk and cherry-pick for 1.1.

langshao22 commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -623,20 +624,10 @@ public boolean commitOffsetsSync(Map<TopicPartition, OffsetAndMetadata> offsets,
-    public void maybeAutoCommitOffsetsNow() {
-        if (autoCommitEnabled && !coordinatorUnknown())
+    public void maybeAutoCommitOffsetsAsync(long now) {
+        if (autoCommitEnabled && now >= nextAutoCommitDeadline) {
```
This modification can lead to repeated commit offset because of nextAutoCommitDeadline has not increased. Could someone tell me this is normal or correct me if I am wrong
